client_it_test.go

```diff
@@ -92,6 +95,163 @@ func TestClientLifecycleEvents(t *testing.T) {
+				atomic.SwapInt32(&lifecycleEventReceived, 0)
+			}
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		assert.NotEqual(t, types.UUID{}, subscriptionID, "subscription UUID should not be empty")
+		if err = client.RemoveLifecycleListener(subscriptionID); err != nil {
+			t.Fatal(err)
+		}
+		if err = client.Shutdown(context.Background()); err != nil {
+			t.Fatal(err)
+		}
+		it.Eventually(t, func() bool {
+			return !client.Running() && lifecycleEventReceived == 1
+		})
+	})
+}
+
+func TestClient_AddMembershipListener(t *testing.T) {
+	var (
+		wgAdded,
+		wgRemoved sync.WaitGroup
+		memberCount int = 3
+	)
+	wgAdded.Add(1)
+	wgRemoved.Add(1)
+	ctx := context.Background()
+	cls := it.StartNewClusterWithOptions("client-subscribed-add-membership-listener", 15701, memberCount)
+	cfg := cls.DefaultConfig()
+	client, err := hz.StartNewClientWithConfig(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func(ctx context.Context, client *hz.Client) {
+		err = client.Shutdown(ctx)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}(ctx, client)
+	subscriptionID, err := client.AddMembershipListener(func(event cluster.MembershipStateChanged) {
+		switch event.State {
+		case cluster.MembershipStateAdded:
+			t.Logf("MembershipStateAdded")
+			wgAdded.Done()
+		case cluster.MembershipStateRemoved:
+			t.Logf("MembershipStateRemoved")
+			wgRemoved.Done()
+		}
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	assert.NotEqual(t, types.UUID{}, subscriptionID, "subscription UUID should not be empty")
+	// member which is not connected by the client initially.
+	uuid := cls.MemberUUIDs[1]
+	//uuid := cls.MemberUUIDs[0]
+	ok, err := cls.RC.TerminateMember(ctx, cls.ClusterID, uuid)
+	it.Must(err)
+	assert.True(t, ok, "rc cannot terminate member")
+	wgRemoved.Wait()
+	_, err = cls.RC.StartMember(ctx, cls.ClusterID)
+	it.Must(err)
+	wgAdded.Wait()
+}
+
+func TestClient_RemoveMembershipListener(t *testing.T) {
+	var added int32 = 1
+	ctx := context.Background()
+	cls := it.StartNewClusterWithOptions("client-subscribed-add-membership-listener", 15701, 2)
+	cfg := cls.DefaultConfig()
+	client, err := hz.StartNewClientWithConfig(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func(ctx context.Context, client *hz.Client) {
+		err = client.Shutdown(ctx)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}(ctx, client)
+	subscriptionID, err := client.AddMembershipListener(func(event cluster.MembershipStateChanged) {
+		switch event.State {
+		case cluster.MembershipStateRemoved:
+			t.Logf("MembershipStateRemoved")
+			atomic.SwapInt32(&added, 0)
+		}
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	assert.NotEqual(t, types.UUID{}, subscriptionID, "subscription UUID should not be empty")
+	err = client.RemoveMembershipListener(subscriptionID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ok, err := cls.RC.TerminateMember(ctx, cls.ClusterID, cls.MemberUUIDs[1])
+	it.Must(err)
+	assert.True(t, ok, "rc cannot terminate member")
+	go func() {
+		time.Sleep(time.Minute * 1)
```

Comment:
Why wait for a minute?

Reply:
`cls.RC.TerminateMember` makes a call to the remote controller, but we don't know when the exact time of the publishing that event on the client side. Therefore I wait for a second, we can decrease is to a few second or so ?

Reply:
It waits for a minute in the code above. Usually sleeping in tests means there's something else wrong. I don't think waiting is necessary here.

Reply:
Fixed, [74c3f9d](https://github.com/hazelcast/hazelcast-go-client/pull/809/commits/74c3f9dd5fc4c20247ec31de821b2bc005d03252).